core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -104,19 +119,40 @@ public void run() {
   /**
    * Transfers all pending and future requests and mutations to the given stream.
    *
-   * <p>No-op if either this method or {@link #cancel} have already been called.
+   * <p>This method must be called at most once.  Extraneous calls will throw and end up cancelling
+   * the given streams.
+   *
+   * <p>If {@link #cancel} has been called, this method will cancel the given stream.
    */
   // When this method returns, passThrough is guaranteed to be true
   final void setStream(ClientStream stream) {
+    ClientStream savedRealStream;
+    Status savedError;
     synchronized (this) {
-      // If realStream != null, then either setStream() or cancel() has been called.
-      if (realStream != null) {
-        return;
+      savedRealStream = realStream;
+      savedError = error;
+      if (savedRealStream == null) {
+        realStream = checkNotNull(stream, "stream");
       }
-      realStream = checkNotNull(stream, "stream");
     }
 
-    drainPendingCalls();
+    if (savedRealStream == null) {
+      drainPendingCalls();
+    } else {
+      // If realStream was not null, then either setStream() or cancel() must had been called,
+      // we will cancel and discard the given stream.
+      // ClientStream.cancel() must be called after start()
+      stream.start(NOOP_STREAM_LISTENER);
+      if (savedError != null) {
+        stream.cancel(savedError);
```

Comment:
> I think the real stream should be cancelled with savedError which was used to cancel the delayed stream.

That probably doesn't matter at all. A simple Status.CANCELLED.withDescription("cancelled before setStream") or "the stream was only started so it could be cancelled" would seem fine to me. The application won't end up seeing the status. When debugging, I would probably prefer to see the hard-coded status instead of `savedError` as well, since the reason for the cancel is the limited API that required the start() in the first place.

> This branch is for calling setStream() after cancel(), which is legit

Hmm... If it is legit, we have to make sure it doesn't happen frequently and could only impact a few streams. And it seems that doesn't hold for `MetadataApplierImpl` and may not hold for `DelayedClientTransport{,2}`. That would be the fix could actually cause stability problems whereas the current code isn't causing any. Hmm...